rln/src/ffi2.rs

```diff
@@ -0,0 +1,1045 @@
+                .iter()
+                .map(|fr| CFr::from(*fr))
+                .collect::<Vec<_>>()
+                .into();
+
+            let path_index: repr_c::Vec<u8> = proof.get_path_index().into();
+
+            let merkle_proof = FFI2_MerkleProof {
+                path_elements,
+                path_index,
+            };
+
+            CResult {
+                ok: Some(Box_::new(merkle_proof)),
+                err: None,
+            }
+        }
+        Err(err) => CResult {
+            ok: None,
+            err: Some(err.to_string().into()),
+        },
+    }
+}
+
+////////////////////////////////////////////////////////
+// zkSNARKs APIs
+////////////////////////////////////////////////////////
+
+#[cfg(not(feature = "stateless"))]
+#[ffi_export]
+pub fn ffi2_generate_rln_proof(
+    rln: &repr_c::Box<FFI2_RLN>,
+    identity_secret: &CFr,
+    user_message_limit: &CFr,
+    message_id: &CFr,
+    x: &CFr,
+    external_nullifier: &CFr,
+    leaf_index: usize,
+) -> CResult<repr_c::Box<FFI2_RLNProof>, repr_c::String> {
+    let proof = match rln.tree.proof(leaf_index) {
+        Ok(proof) => proof,
+        Err(err) => {
+            return CResult {
+                ok: None,
+                err: Some(err.to_string().into()),
+            };
+        }
+    };
+
+    let path_elements: Vec<Fr> = proof.get_path_elements();
+    let identity_path_index: Vec<u8> = proof.get_path_index();
+
+    let mut identity_secret = identity_secret.0;
+    let rln_witness = RLNWitnessInput {
+        identity_secret: IdSecret::from(&mut identity_secret),
+        user_message_limit: user_message_limit.0,
+        message_id: message_id.0,
+        path_elements,
+        identity_path_index,
+        x: x.0,
+        external_nullifier: external_nullifier.0,
+    };
+
+    let proof_values = match proof_values_from_witness(&rln_witness) {
+        Ok(pv) => pv,
+        Err(err) => {
+            return CResult {
+                ok: None,
+                err: Some(err.to_string().into()),
+            };
+        }
+    };
+
+    let proof = match generate_proof(&rln.proving_key, &rln_witness, &rln.graph_data) {
+        Ok(proof) => proof,
+        Err(err) => {
+            return CResult {
+                ok: None,
+                err: Some(err.to_string().into()),
+            };
+        }
+    };
+
+    CResult {
+        ok: Some(Box_::new(FFI2_RLNProof {
+            proof_values,
+            proof,
+        })),
+        err: None,
+    }
+}
+
+#[cfg(not(feature = "stateless"))]
+#[ffi_export]
+pub fn ffi2_verify_rln_proof(
+    rln: &repr_c::Box<FFI2_RLN>,
+    proof: &repr_c::Box<FFI2_RLNProof>,
+) -> CResult<repr_c::Box<bool>, repr_c::String> {
+    // Verify the proof
+    match verify_proof(&rln.proving_key.0.vk, &proof.proof, &proof.proof_values) {
+        Ok(proof_verified) => {
+            // Verify the root
+            let roots_verified = rln.tree.root() == proof.proof_values.root;
```

Comment:
So where is the message itself? Your input should be proof_data and signal and then you also do check, that proof_data.x == signal